nomt/src/beatree/ops/bit_ops.rs

```diff
@@ -64,64 +64,141 @@ pub fn reconstruct_key(maybe_prefix: Option<RawPrefix>, separator: RawSeparator)
     let (separator_bytes, separator_bit_init, separator_bit_len) = separator;
 
     // where the separator will start to be stored
-    let mut key_offset = match prefix_byte_len {
+    let init_destination = match prefix_byte_len {
         0 => 0,
         len if prefix_end_bit_offset == 0 => len,
         // overlap between the end of the prefix and the beginning of the separator
         len => len - 1,
     };
 
+    // shift the separator and store it in the key
+    bitwise_memcpy(
+        &mut key[init_destination..],
+        prefix_end_bit_offset,
+        separator_bytes,
+        separator_bit_init,
+        separator_bit_len,
+    );
+
+    if prefix_byte_len != 0 {
+        // UNWRAP: prefix_byte_len can be different than 0 only if maybe_prefix is Some
+        let prefix = maybe_prefix.unwrap();
+
+        // copy the prefix into the key up to the penultimate byte
+        key[0..prefix_byte_len - 1].copy_from_slice(&prefix.0[0..prefix_byte_len - 1]);
+
+        // copy the last byte of the prefix without interfering with the separator
+        let mask_shift = 8 - (prefix_bit_len % 8) as u32;
+        let mask = 1u8.checked_shl(mask_shift).map(|m| !(m - 1)).unwrap_or(255);
+        key[prefix_byte_len - 1] |= prefix.0[prefix_byte_len - 1] & mask;
+    }
+
+    key
+}
+
+fn first_chunk_mask(bit_init: usize) -> u64 {
+    let mask_shift = (7 - bit_init) as u32 + 1 + (8 * 7);
+    1u64.checked_shl(mask_shift)
+        .map(|m| m - 1)
+        .unwrap_or(u64::MAX)
+}
+
+fn last_chunk_mask(bit_init: usize, bit_len: usize, n_chunks: usize) -> u64 {
+    let unused_last_bits = (bit_init + bit_len).saturating_sub((n_chunks - 1) * 64);
+    // saturating_sub is necessary to prevent overflow in the rare case in which a right
+    // shift may shift some bits onto the `n_chunks` chunk, leading to more than 64 bits
+    // on the `n_chunks - 1` chunk.
+    1u64.checked_shl(64u32.saturating_sub(unused_last_bits as u32))
+        .map(|m| !(m - 1))
+        .unwrap_or(0)
+}
+
+/// Copy `source_bit_len` from the source bytes, starting from the `source_bit_init`ith bit
+/// within the first byte into the destination, starting from the `destination_bit_init`ith bit.
+///
+/// `destination` must be long enough to store the bits, accounting also possible shifts.
+/// `source` must have a length multiple of 8 bytes.
+///
+/// All the bits in `destination` not involved in the copy will be left unchanged.
+pub fn bitwise_memcpy(
+    destination: &mut [u8],
+    destination_bit_init: usize,
+    source: &[u8],
+    source_bit_init: usize,
```

Comment:
style: a slice start is usually referred to as "start" not "init". "init" implies that it's a value, not a position.

Reply:
Thanks, yes, this makes perfect sense. I generally use them interchangeably, but now that I have a clearer idea, I will use them properly!